source/auth/tests/mongodb-oidc.rst

```diff
@@ -4,164 +4,397 @@ MongoDB OIDC
+- Clear the cache.
+- Create a client that uses the OIDC url and a request callback, and an
+  ``ALLOWED_HOSTS`` that is an empty list.
+- Assert that a ``find`` operation fails with a client-side error.
+- Close the client.
+- Create a client that uses the OIDC url and a request callback, and an
+  ``ALLOWED_HOSTS`` that contains ["localhost1"].
+- Assert that a ``find`` operation fails with a client-side error.
+- Close the client.
+
+AWS Automatic Auth
+==================
 
-#. Create a request callback that reads in the generated ``test_user1`` token
-file.
-#. Create a client with a url of the form  ``mongodb://test_user1@localhost/?authMechanism=MONGODB-OIDC`` and the OIDC request callback.
-#. Perform a ``find`` operation.
-#. Clear the cache.
+Drivers MUST be able to authenticate using the "aws" provider workflow
+simulating an EC2 instance with an enabled web identity token provider,
+generated by Drivers Evergreen Tools.
+
+Single Principal
+~~~~~~~~~~~~~~~~
+- Create a client with a url of the form ``mongodb://localhost/?authMechanism=MONGODB-OIDC&authMechanismProperties=PROVIDER_NAME:aws``.
+- Perform a ``find`` operation that succeeds.
+- Close the client.
+
+Multiple Principal User 1
+~~~~~~~~~~~~~~~~~~~~~~~~~
+- Create a client with a url of the form ``mongodb://localhost:27018/?authMechanism=MONGODB-OIDC&authMechanismProperties=PROVIDER_NAME:aws&directConnection=true&readPreference=secondaryPreferred``.
+- Perform a ``find`` operation that succeeds.
+- Close the client.
+
+Multiple Principal User 2
+~~~~~~~~~~~~~~~~~~~~~~~~~
+- Set the ``AWS_WEB_IDENTITY_TOKEN_FILE`` environment variable
+  to the location of valid ``test_user2`` credentials.
+- Create a client with a url of the form ``mongodb://localhost:27018/?authMechanism=MONGODB-OIDC&authMechanismProperties=PROVIDER_NAME:aws&directConnection=true&readPreference=secondaryPreferred``.
+- Perform a ``find`` operation that succeeds.
+- Close the client.
+- Restore the ``AWS_WEB_IDENTITY_TOKEN_FILE`` environment variable
+  to the location of valid ``test_user2`` credentials.
+
+Allowed Hosts Ignored
+~~~~~~~~~~~~~~~~~~~~~
+- Create a client with a url of the form ``mongodb://localhost/?authMechanism=MONGODB-OIDC&authMechanismProperties=PROVIDER_NAME:aws``, and an
+  ``ALLOWED_HOSTS`` that is an empty list.
+- Assert that a ``find`` operation succeeds.
+- Close the client.
+
+Callback Validation
+===================
 
+Valid Callbacks
+~~~~~~~~~~~~~~~
+- Clear the cache.
+- Create request and refresh callback that validate their inputs and return
+  a valid token.  The request callback must return a token that expires in
+  one minute.
+- Create a client that uses the above callbacks.
+- Perform a ``find`` operation that succeeds.  Verify that the request
+  callback was called with the appropriate inputs, including the timeout
+  parameter if possible.  Ensure that there are no unexpected fields.
+- Perform another ``find`` operation that succeeds.  Verify that the refresh
```

Comment:
How is the refresh callback supposed to be called in this test? Is it expected that the second `find` happens on a separate connection in order for another initial handshake to occur? Should this be behaving as later tests that `find` first, then close the client before creating a new client and then issuing a new `find`?

Reply:
Edit: I seemed to answer my question looking at the Python PR - so could we update these 2 areas to note the client must be closed and a new one created like the tests that do the same later in the prose tests?

Reply:
Updated